### PR TITLE
Make module types explicit and convert some to ESM

### PR DESCRIPTION
**Pull request by steverep**
## Proposed change

These changes make the repository compliant with the rules for [determining the module system](https://nodejs.org/api/packages.html#determining-module-system) - ESM or CommonJS.  The root change is to specify `"type": "module"` in `package.json` so that `.js` and `.ts` files are always interpreted as ESM, and rename any CommonJS modules to `.cjs` or `.cts` (obviously just the build scripts and tool config files here).

The impetus for this is that most of our build tools comply with these rules and even take configuration cues from it.  The Webpack build actually produces slightly different code with these changes (not 100% clear on why just yet, but it works fine).

Here are some more details:
- Everything in `build-scripts` was simply renamed to use the `.cjs` extension, and correcting the `require` statements accordingly.
- A single change was made to the Webpack config to set `resolver.fullySpecified: false` (just tells Webpack to ignore that ESM rule so we can continue leaving off extensions on imports).
- For `*.config.js` files, I converted most to ESM with the exception of the root Webpack config because it's used by the ESLint resolver plugin which doesn't support ESM yet
- For `gulpfile.js`, I abandoned the `require-dir` package which is dependent on some deprecated `node` features.  I just rewrote it using a glob and loop.
- See my review note regarding `hassio/config.js`

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: hassio/config.cjs
Comment (steverep) [inline hassio/config.cjs:1]: Where is this config used?  I searched and couldn't find it imported anywhere.

**Comment by bramkragten** [comment]
Nice! :tada:

**Comment by steverep** [comment]
I've got a few more tricks to apply after this one merges... 🪄

**Comment by steverep** [comment]
Added a small related change to TS config to set `esModuleInterop: true` (mostly doesn't affect anything but more correct)

**Review by bramkragten — APPROVED** [review]

File: hassio/config.cjs
Comment (steverep) [reply hassio/config.cjs:1]: @bramkragten so is this file defunct?  If not, I probably broke something by renaming.

**Pull request merged**